### PR TITLE
Added DealWheel View as library incl. sample app [replacing old version]

### PR DESCRIPTION
**Pull request by Informaticore**
This is the new implementation of the DealWheel View. It is placed inside a library and used in a sample App of the same project. 
I replaced the whole implementation of Resul (canvas drawing).

### CODE REVIEW

File: app/src/main/res/mipmap-anydpi-v26/ic_launcher.xml [hunk 1]
```diff
@@ -1,5 +1,5 @@
 <?xml version="1.0" encoding="utf-8"?>
 <adaptive-icon xmlns:android="http://schemas.android.com/apk/res/android">
-    <background android:drawable="@drawable/ic_launcher_background" />
-    <foreground android:drawable="@mipmap/ic_launcher_foreground" />
+    <background android:drawable="@drawable/ic_launcher_background"/>
+    <foreground android:drawable="@drawable/ic_launcher_foreground"/>
```
Comment (gerokeller) [inline app/src/main/res/mipmap-anydpi-v26/ic_launcher.xml]: Why not mipmap?

File: app/src/main/res/mipmap-anydpi-v26/ic_launcher_round.xml [hunk 2]
```diff
@@ -1,5 +1,5 @@
 <?xml version="1.0" encoding="utf-8"?>
 <adaptive-icon xmlns:android="http://schemas.android.com/apk/res/android">
-    <background android:drawable="@drawable/ic_launcher_background" />
-    <foreground android:drawable="@mipmap/ic_launcher_foreground" />
+    <background android:drawable="@drawable/ic_launcher_background"/>
+    <foreground android:drawable="@drawable/ic_launcher_foreground"/>
```
Comment (gerokeller) [inline app/src/main/res/mipmap-anydpi-v26/ic_launcher_round.xml]: Why not mipmap?

File: lib/src/main/java/com/crvsh/dealwheellibrary/Deal.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
+package com.crvsh.dealwheellibrary;
+
+public class Deal {
+
+
```
Comment (gerokeller) [inline lib/src/main/java/com/crvsh/dealwheellibrary/Deal.java]: Quite a lot of whitespace here

File: lib/src/main/java/com/crvsh/dealwheellibrary/DealWheelFragment.kt [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+import android.view.LayoutInflater
+import android.view.View
+import android.view.ViewGroup
+import kotlinx.android.synthetic.main.deal_wheel_fragment.*
+import java.util.*
+
+class DealWheelFragment : Fragment() {
+
+    companion object {
+        val BUNDLE_DEALS_JSON: String = "dealsJson"
```
Comment (gerokeller) [inline lib/src/main/java/com/crvsh/dealwheellibrary/DealWheelFragment.kt]: Why is this needed? It doesn't seem to be used in `DealWheelFragment`

File: lib/src/main/java/com/crvsh/dealwheellibrary/WheelAnimationListener.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    public void onAnimationUpdate(ValueAnimator valueAnimator) {
+        float currentRotation = (float) valueAnimator.getAnimatedValue();
+        if ((int) Math.abs(currentRotation) / 30 == plings) {
+            onTick();
+            soundPool.playWheelSpin();
+            plings++;
+        }
+    }
+
+    public void onTick() {
```
Comment (gerokeller) [inline lib/src/main/java/com/crvsh/dealwheellibrary/WheelAnimationListener.java]: Is this needed?

File: lib/src/main/java/com/crvsh/dealwheellibrary/WheelAnimationListener.java [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+
+    }
+
+    @Override
+    public void onAnimationStart(Animator animator) {
+
```
Comment (gerokeller) [inline lib/src/main/java/com/crvsh/dealwheellibrary/WheelAnimationListener.java]: Whitespace in empty overriden method?

File: lib/src/main/java/com/crvsh/dealwheellibrary/WheelSoundPool.java [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+            AudioAttributes audioAttrib = new AudioAttributes.Builder()
+                    .setUsage(AudioAttributes.USAGE_GAME)
+                    .setContentType(AudioAttributes.CONTENT_TYPE_SONIFICATION)
+                    .build();
+
+            SoundPool.Builder builder = new SoundPool.Builder();
+            builder.setAudioAttributes(audioAttrib).setMaxStreams(MAX_STREAMS);
+            soundPool = builder.build();
+        } else {
+            soundPool = new SoundPool(MAX_STREAMS, AudioManager.STREAM_MUSIC, 0);
```
Comment (gerokeller) [inline lib/src/main/java/com/crvsh/dealwheellibrary/WheelSoundPool.java]: That else could be removed and `soundPool = new SoundPool(MAX_STREAMS, AudioManager.STREAM_MUSIC, 0);` could be moved before the check for the `SDK_INT`.

File: lib/src/main/java/com/crvsh/dealwheellibrary/WheelView.java [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+import android.widget.RelativeLayout;
+import android.widget.TextView;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class WheelView extends FrameLayout {
+    public static final int DEAL_ITEMS_MIN_COUNT = 12;
+    public static final float WHEEL_ROTATION_5_TIMES = 1800f;
+    private final GestureDetector gestureDetector;
```
Comment (gerokeller) [inline lib/src/main/java/com/crvsh/dealwheellibrary/WheelView.java]: Whitespace between `static` and normal properties?

File: lib/src/main/java/com/crvsh/dealwheellibrary/WheelView.java [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+            dealItem.setLayoutParams(layoutParams);
+            TextView dealName = dealItem.findViewById(R.id.deal_name);
+            TextView dealDiscount = dealItem.findViewById(R.id.deal_discount);
+            TextView dealIndex = dealItem.findViewById(R.id.deal_index);
+
+            dealName.setText(deals.get(i % deals.size()).getName());
+            dealDiscount.setText(deals.get(i % deals.size()).getDiscount());
+
+            dealIndex.setText(String.format("%02d", ((i % deals.size()) + 1)));
+//            dealIndex.setText("12");
```
Comment (gerokeller) [inline lib/src/main/java/com/crvsh/dealwheellibrary/WheelView.java]: Can this be removed?

File: lib/src/main/java/com/crvsh/dealwheellibrary/WheelView.java [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+    public boolean onTouchEvent(MotionEvent event) {
+        if (animationRunning) {
+            return false;
+        }
+
+        if (event.getAction() == MotionEvent.ACTION_DOWN) {
+            yDown = event.getY();
+        } else if (event.getAction() == MotionEvent.ACTION_UP) {
+            isDirectionUp = yDown > event.getY();
+
```
Comment (gerokeller) [inline lib/src/main/java/com/crvsh/dealwheellibrary/WheelView.java]: Whitespace?

File: lib/src/main/java/com/crvsh/dealwheellibrary/WheelView.java [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+    @Override
+    protected void onSizeChanged(int w, int h, int oldw, int oldh) {
+        super.onSizeChanged(w, h, oldw, oldh);
+        centerX = w / 2;
+        centerY = h / 2;
+
+        frameLayout.setTranslationX(-(wheelItemCircleSize * 0.57f));
+        frameLayout.setTranslationY(-(wheelItemCircleSize / 2 - h / 2));
+
```
Comment (gerokeller) [inline lib/src/main/java/com/crvsh/dealwheellibrary/WheelView.java]: Whitespace

File: lib/src/main/java/com/crvsh/dealwheellibrary/WheelView.java [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+                        bounceAnimator.setInterpolator(new OvershootInterpolator());
+                        bounceAnimator.addUpdateListener(this);
+                        bounceAnimator.addListener(this);
+                        bounceAnimator.start();
+                    }
+                }
+
+                @Override
+                public void onTick() {
+//                    if (isDirectionUp) {
```
Comment (gerokeller) [inline lib/src/main/java/com/crvsh/dealwheellibrary/WheelView.java]: Should we keep those commented lines in the code base?

File: lib/src/main/res/drawable/wheel_background.xml [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+        <layer-list xmlns:android="http://schemas.android.com/apk/res/android">
+            <item>
+                <shape xmlns:android="http://schemas.android.com/apk/res/android"
+                       android:shape="oval">
+                    <solid android:color="#eeeeee"/>
+                </shape>
+            </item>
+
+            <item
+                    android:bottom="@dimen/circle_inner_padding"
```
Comment (gerokeller) [inline lib/src/main/res/drawable/wheel_background.xml]: For `item` the properties start on the next line for other on the same line. Should this be consistent?

File: lib/src/main/res/layout/deal_wheel_item.xml [hunk 14]
```diff
@@ -0,0 +1,57 @@
+<?xml version="1.0" encoding="utf-8"?>
+<RelativeLayout xmlns:android="http://schemas.android.com/apk/res/android"
+                xmlns:tools="http://schemas.android.com/tools"
+                android:layout_width="wrap_content"
+                android:layout_height="wrap_content">
+
+    <LinearLayout
+            android:id="@+id/index_container"
```
Comment (gerokeller) [inline lib/src/main/res/layout/deal_wheel_item.xml]: For some xml items the properties start on a new line for other on the same line. Should this be consistent?

**Comment by gerokeller** [comment]
This project needs linting and travis setup. Can someone please add this as an issue on the project? @Informaticore @luserdroid

File: app/src/main/res/mipmap-anydpi-v26/ic_launcher.xml (hunk 1 above)
Comment (Informaticore) [reply app/src/main/res/mipmap-anydpi-v26/ic_launcher.xml]: its a change of the new library project I created and it is just for the sample app so: I don't care

File: lib/src/main/java/com/crvsh/dealwheellibrary/Deal.java (hunk 3 above)
Comment (Informaticore) [reply lib/src/main/java/com/crvsh/dealwheellibrary/Deal.java]: Gets deleted anyway - was about to create a new class with kotlin for the deals

File: lib/src/main/java/com/crvsh/dealwheellibrary/DealWheelFragment.kt (hunk 4 above)
Comment (Informaticore) [reply lib/src/main/java/com/crvsh/dealwheellibrary/DealWheelFragment.kt]: but it will be

File: lib/src/main/java/com/crvsh/dealwheellibrary/WheelAnimationListener.java (hunk 5 above)
Comment (Informaticore) [reply lib/src/main/java/com/crvsh/dealwheellibrary/WheelAnimationListener.java]: yes sir

File: lib/src/main/java/com/crvsh/dealwheellibrary/WheelView.java (hunk 9 above)
Comment (Informaticore) [reply lib/src/main/java/com/crvsh/dealwheellibrary/WheelView.java]: done

File: lib/src/main/java/com/crvsh/dealwheellibrary/WheelView.java (hunk 10 above)
Comment (Informaticore) [reply lib/src/main/java/com/crvsh/dealwheellibrary/WheelView.java]: someone likes to say "whitespaces"

File: lib/src/main/java/com/crvsh/dealwheellibrary/WheelView.java (hunk 12 above)
Comment (Informaticore) [reply lib/src/main/java/com/crvsh/dealwheellibrary/WheelView.java]: yes sir because they never got committed and they are important later

File: lib/src/main/res/drawable/wheel_background.xml (hunk 13 above)
Comment (Informaticore) [reply lib/src/main/res/drawable/wheel_background.xml]: yes it should and is part of the formatting - I don't want to change it or discous this now. Thats part of the "how to apply formatting to the project" topic

File: lib/src/main/java/com/crvsh/dealwheellibrary/WheelView.java (hunk 11 above)
Comment (Informaticore) [reply lib/src/main/java/com/crvsh/dealwheellibrary/WheelView.java]: you love this stuff. 
I have the feeling whitespacing shouldn't be a topic. Especially if it is about a whitespace at the end of a method. We can have a looong discoussin about this or we just ignore it because it will not have a real effect on the readability or the quality of the code. 
My formatting is reducing whitespacing to at least be 2 (if it is more then 2 lines).

File: lib/src/main/java/com/crvsh/dealwheellibrary/WheelSoundPool.java (hunk 7 above)
Comment (Informaticore) [reply lib/src/main/java/com/crvsh/dealwheellibrary/WheelSoundPool.java]: hmm.. no, how come? the one is using new the otherone is using the builder. huge difference.
If I put it before the if it will be declared like this ll the time and gets overridden by the builder if the IF kicks in. I more like it this way so it is clear whats happening.

File: lib/src/main/java/com/crvsh/dealwheellibrary/DealWheelFragment.kt (hunk 4 above)
Comment (gerokeller) [reply lib/src/main/java/com/crvsh/dealwheellibrary/DealWheelFragment.kt]: In my opinion we should only add code to the codebase that *is* being used

File: lib/src/main/java/com/crvsh/dealwheellibrary/WheelAnimationListener.java (hunk 5 above)
Comment (gerokeller) [reply lib/src/main/java/com/crvsh/dealwheellibrary/WheelAnimationListener.java]: Can you elaborate? This is a public method that does nothing and is not overriding anything. I think I'm missing something here.

File: lib/src/main/java/com/crvsh/dealwheellibrary/WheelView.java (hunk 11 above)
Comment (gerokeller) [reply lib/src/main/java/com/crvsh/dealwheellibrary/WheelView.java]: I think it should be consistent. Sometimes there is whitespace before the closing bracket of and else sometimes and the end of the initial if, sometime and the end of a method and sometimes at the beginning of the method. There should be one consistent way how we use whitespace to separate logical parts of the code.
I think whitespace is important and something everyone should pay attention to. You are training your eyes to follow certain patterns, every time this is surprising you are doing double cycles to understand what is going on.

File: lib/src/main/java/com/crvsh/dealwheellibrary/WheelView.java (hunk 12 above)
Comment (gerokeller) [reply lib/src/main/java/com/crvsh/dealwheellibrary/WheelView.java]: Yes, they just got committed. If this is some work that you want to work on later, stash it, create a fork and push it there. Commented lines just to keep a reference of code that is unfinished is an anti pattern in my opinion.

File: lib/src/main/res/drawable/wheel_background.xml (hunk 13 above)
Comment (gerokeller) [reply lib/src/main/res/drawable/wheel_background.xml]: Are you keeping track of the topic? When is this going to be discussed?

File: app/src/main/res/mipmap-anydpi-v26/ic_launcher.xml (hunk 1 above)
Comment (gerokeller) [reply app/src/main/res/mipmap-anydpi-v26/ic_launcher.xml]: Makes me sad. So little love

File: lib/src/main/java/com/crvsh/dealwheellibrary/DealWheelFragment.kt (hunk 4 above)
Comment (Informaticore) [reply lib/src/main/java/com/crvsh/dealwheellibrary/DealWheelFragment.kt]: thats a special case... I already started te next topic but I dont want to remove it just for the commit :) its a dilemma

File: lib/src/main/java/com/crvsh/dealwheellibrary/WheelAnimationListener.java (hunk 5 above)
Comment (Informaticore) [reply lib/src/main/java/com/crvsh/dealwheellibrary/WheelAnimationListener.java]: yes you do... it can be used... maybe ;)
I was just leaving it there so other devs can see that this exists.

File: lib/src/main/res/layout/deal_wheel_item.xml (hunk 14 above)
Comment (Informaticore) [reply lib/src/main/res/layout/deal_wheel_item.xml]: I am gone for a week now but it can be a topic for the tech forum

**Comment by gerokeller** [comment]
Moving on with this PR because I believe non of discussion items are urgent or important.
